ddprof-lib/src/main/cpp/rustDemangler.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2021, 2023 Datadog, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include "rustDemangler.h"
+
+#include <array>
+
+namespace RustDemangler {
+
+// With some exceptions we don't handle here, v0 Rust symbols can end in a
+// prefix followed by a 16-hexdigit hash, which must be removed
+const std::string hash_pre = "::h";
+const std::string hash_eg = "0123456789abcdef";
+
+const std::array<std::pair<std::string, std::string>, 9> patterns = {{
+    {"..", "::"},
+    {"$C$", ","},
+    {"$BP$", "*"},
+    {"$GT$", ">"},
+    {"$LT$", "<"},
+    {"$LP$", "("},
+    {"$RP$", ")"},
+    {"$RF$", "&"},
+    {"$SP$", "@"},
+}};
+
+inline bool is_hexdig(const char c) {
+  return (c >= 'a' && c <= 'f') || (c >= '0' && c <= '9');
+}
+
+// Simple conversion from hex digit to integer
+inline int hex_to_int(char dig) {
+  constexpr int k_a_hex_value = 0xa;
+  if (dig >= '0' && dig <= '9') {
+    return dig - '0';
+  }
+  if (dig >= 'a' && dig <= 'f') {
+    return dig - 'a' + k_a_hex_value;
+  }
+  if (dig >= 'A' && dig <= 'F') {
+    return dig - 'A' + k_a_hex_value;
+  }
+  return -1;
+}
+
+// Minimal check that a string can end, and does end, in a hashlike substring
+inline bool has_hash(const std::string &str) {
+  // If the size can't conform, then the string is invalid
+  if (str.size() <= hash_pre.size() + hash_eg.size()) {
```

Comment:
Will the calls to `size()` be somehow constant folded when the string is `const`? 
I am genuinely curious here - I guess it should be a pretty simple optimization to be done by compiler, but I have no clue.

Reply:
Ah, right--it's worth thinking about this a bit.  I adapted this code from a system which was already using `std::string` everywhere, so there might be some unexpected surprises.

Roughly speaking,
* `str.size()` is a struct getter against the string's `size` parameter, which was computed at time of creation
* `hash_pre.size()` and `hash_eg.size()` are the same, except the scan was a one-time thing.  In later C++ standards, we could have used `constexpr std::string_view` to move even that scan to compile-time.

In this case, the size of the string is computed at time of creation.  Ultimately, `str` is created from a C string, and I create it the lazy way (line 89 of flightRecorder.cpp), so there's an implicit `strlen()` that happens (as well as a copy).  Constructing the string with the length as a parameter avoids the `strlen()`, but the `memcpy()` is still linear (albeit "fast", probably).

In later C++ standards, there'a `std::string_view` which can be used to create a non-owning "view" into a string (be it a C++ std::string or a C-style char * equivalent).  This is one way of avoiding any kinds of linear penalties, but it isn't compatible with the project (we'd need to build in C++17 mode, I think).



Anyway, my point is that your observation is indeed true for the consts, but we should be wary of the fact that `str` did require a `strlen()` and a `memcpy()` in order to be created.